consensus/src/lib.rs

```diff
@@ -97,6 +100,55 @@ cfg_if::cfg_if! {
             ) -> impl Future<Output = oneshot::Receiver<bool>> + Send;
         }
 
+        /// Application is the interface responsible for building new blocks on top of consensus-provided parent
+        /// commitments as well as receiving finalized updates from [crate::marshal].
+        pub trait Application<E>: Clone + Send + 'static
+        where
+            E: Rng + Spawner + Metrics + Clock
+        {
+            /// Context is metadata provided by the consensus engine associated with a given payload.
+            ///
+            /// This often includes things like the proposer, view number, the height, or the epoch.
+            type Context: Epochable;
+
+            /// The block type produced by the application's builder.
+            type Block: Block;
+
+            /// Payload used to initialize the consensus engine in the first epoch.
+            fn genesis(&mut self) -> impl Future<Output = Self::Block> + Send;
+
+            /// Build a new block on top of the provided parent commitment. If the build job fails,
+            /// returns [None].
+            fn build(
+                &mut self,
+                r_context: E,
+                context: Self::Context,
+            ) -> impl Future<Output = Option<Self::Block>> + Send;
```

Comment:
It would be nice if propose still retained some idea of when the proposal was no longer required?